theano/gpuarray/opt.py

```diff
@@ -570,22 +854,29 @@ def local_gpu_pdbbreakpoint_op(node):
 
 @register_opt('fast_compile')
 @op_lifter([IfElse])
-def local_gpua_lazy_ifelse(node, context_name):
-    if node.op.gpu:
+@register_opt2([IfElse], 'fast_compile')
+def local_gpua_lazy_ifelse(op, context_name, inputs, outputs):
+    if op.gpu:
         return
-    c = node.inputs[0]
+    # this node is already on GPU, so don't change the graph
+    if isinstance(inputs[0].type, GpuArrayType):
+        return
```

Comment:
Wait, we should discuss that tomorrow.
The important thing is that, when called from the new optimizer (where `@register_opt2` puts it), this `if` is not actually checking that the node is already on the GPU. Even if we wanted to be strict at that point, this would not be the right check (and I think it would not be the right place to check).